File: android/app/src/main/java/com/move/GetCurrentLocation.kt
// GetCurrentLocation.kt
package com.solagroup.bandmate

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.app.ActivityCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class GetCurrentLocation(private val reactContext: ReactApplicationContext) : 
    ReactContextBaseJavaModule(reactContext), LocationListener {
    
    private var locationManager: LocationManager? = null
    private var isLocationUpdatesStarted = false
    private val handler = Handler(Looper.getMainLooper())
    private var timeoutRunnable: Runnable? = null
    private var currentPromise: Promise? = null
    private var gpsLocationListener: LocationListener? = null
    private var networkLocationListener: LocationListener? = null
    
    companion object {
        private const val TAG = "GetCurrentLocation"
        private const val TIMEOUT_DURATION = 15000L // 15 seconds
        private const val MIN_TIME_BETWEEN_UPDATES = 100L // 100ms
        private const val MIN_DISTANCE_CHANGE_FOR_UPDATES = 0f // 0 meters for immediate response
        private const val LAST_KNOWN_LOCATION_MAX_AGE = 5 * 60 * 1000L // 5 minutes
    }

    override fun getName(): String {
        return "GetCurrentLocation"
    }

    @ReactMethod
    fun getCurrentLocation(promise: Promise) {
        // Prevent multiple simultaneous requests
        if (isLocationUpdatesStarted) {
            promise.reject("LOCATION_REQUEST_IN_PROGRESS", "A location request is already in progress")
            return
        }

        if (!hasLocationPermission()) {
            promise.reject("PERMISSION_DENIED", "Location permission not granted")
            return
        }

        try {
            locationManager = reactContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            
            // Check if location services are enabled
            if (!isLocationEnabled()) {
                promise.reject("LOCATION_DISABLED", "Location services are disabled")
                return
            }

            currentPromise = promise
            startLocationUpdates()
            
        } catch (e: Exception) {
            Log.e(TAG, "Error getting location: ${e.message}")
            promise.reject("LOCATION_ERROR", e.message)
        }
    }

    private fun startLocationUpdates() {
        try {
            isLocationUpdatesStarted = true
            
            // Set timeout
            timeoutRunnable = Runnable {
                handleTimeout()
            }
            handler.postDelayed(timeoutRunnable!!, TIMEOUT_DURATION)
            
            // First try to get last known location
            getLastKnownLocation()?.let { location ->
                val locationAge = System.currentTimeMillis() - location.time
                if (locationAge <= LAST_KNOWN_LOCATION_MAX_AGE) {
                    Log.d(TAG, "Using last known location (age: ${locationAge}ms)")
                    handleLocationResult(location)
                    return
                }
            }
            
            // Create location listeners
            gpsLocationListener = object : LocationListener {
                override fun onLocationChanged(location: Location) {
                    Log.d(TAG, "GPS location received: ${location.latitude}, ${location.longitude}")
                    handleLocationResult(location)
                }
                
                override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {}
                override fun onProviderEnabled(provider: String) {}
                override fun onProviderDisabled(provider: String) {}
            }
            
            networkLocationListener = object : LocationListener {
                override fun onLocationChanged(location: Location) {
                    Log.d(TAG, "Network location received: ${location.latitude}, ${location.longitude}")
                    handleLocationResult(location)
                }
                
                override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {}
                override fun onProviderEnabled(provider: String) {}
                override fun onProviderDisabled(provider: String) {}
            }
            
            // Request location updates from both providers
            if (locationManager?.isProviderEnabled(LocationManager.GPS_PROVIDER) == true) {
                locationManager?.requestLocationUpdates(
                    LocationManager.GPS_PROVIDER,
                    MIN_TIME_BETWEEN_UPDATES,
                    MIN_DISTANCE_CHANGE_FOR_UPDATES,
                    gpsLocationListener!!
                )
                Log.d(TAG, "GPS location updates requested")
            }
            
            if (locationManager?.isProviderEnabled(LocationManager.NETWORK_PROVIDER) == true) {
                locationManager?.requestLocationUpdates(
                    LocationManager.NETWORK_PROVIDER,
                    MIN_TIME_BETWEEN_UPDATES,
                    MIN_DISTANCE_CHANGE_FOR_UPDATES,
                    networkLocationListener!!
                )
                Log.d(TAG, "Network location updates requested")
            }
            
        } catch (e: SecurityException) {
            Log.e(TAG, "Security exception: ${e.message}")
            currentPromise?.reject("PERMISSION_DENIED", "Location permission denied")
            cleanup()
        }
    }

    private fun handleLocationResult(location: Location) {
        if (isLocationUpdatesStarted && currentPromise != null) {
            Log.d(TAG, "Location result: lat=${location.latitude}, lon=${location.longitude}, accuracy=${location.accuracy}")
            
            val locationData = WritableNativeMap().apply {
                putDouble("latitude", location.latitude)
                putDouble("longitude", location.longitude)
                putDouble("accuracy", location.accuracy.toDouble())
                putDouble("altitude", if (location.hasAltitude()) location.altitude else 0.0)
                putDouble("speed", if (location.hasSpeed()) location.speed.toDouble() else 0.0)
                putDouble("heading", if (location.hasBearing()) location.bearing.toDouble() else 0.0)
                putDouble("timestamp", location.time.toDouble())
                putString("provider", location.provider ?: "unknown")
            }
            
            currentPromise?.resolve(locationData)
            
            // Emit event for any listeners
            try {
                reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit("locationUpdate", locationData)
            } catch (e: Exception) {
                Log.w(TAG, "Could not emit location event: ${e.message}")
            }
            
            cleanup()
        }
    }

    private fun handleTimeout() {
        if (isLocationUpdatesStarted && currentPromise != null) {
            Log.w(TAG, "Location request timed out")
            
            // Try one more time with last known location
            getLastKnownLocation()?.let { location ->
                Log.d(TAG, "Using stale last known location due to timeout")
                handleLocationResult(location)
                return
            }
            
            currentPromise?.reject("TIMEOUT", "Location request timed out")
            cleanup()
        }
    }

    private fun cleanup() {
        isLocationUpdatesStarted = false
        currentPromise = null
        
        timeoutRunnable?.let { 
            handler.removeCallbacks(it)
            timeoutRunnable = null
        }
        
        try {
            gpsLocationListener?.let { locationManager?.removeUpdates(it) }
            networkLocationListener?.let { locationManager?.removeUpdates(it) }
        } catch (e: SecurityException) {
            Log.e(TAG, "Error removing location updates: ${e.message}")
        }
        
        gpsLocationListener = null
        networkLocationListener = null
    }

    private fun getLastKnownLocation(): Location? {
        if (!hasLocationPermission()) return null
        
        var bestLocation: Location? = null
        
        try {
            val gpsLocation = locationManager?.getLastKnownLocation(LocationManager.GPS_PROVIDER)
            val networkLocation = locationManager?.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
            
            bestLocation = when {
                gpsLocation != null && networkLocation != null -> {
                    // Choose more recent location, or more accurate if times are close
                    val timeDiff = Math.abs(gpsLocation.time - networkLocation.time)
                    if (timeDiff < 60000) { // If within 1 minute, choose more accurate
                        if (gpsLocation.accuracy < networkLocation.accuracy) gpsLocation else networkLocation
                    } else {
                        if (gpsLocation.time > networkLocation.time) gpsLocation else networkLocation
                    }
                }
                gpsLocation != null -> gpsLocation
                networkLocation != null -> networkLocation
                else -> null
            }
            
            bestLocation?.let {
                Log.d(TAG, "Last known location: provider=${it.provider}, accuracy=${it.accuracy}, age=${System.currentTimeMillis() - it.time}ms")
            }
            
        } catch (e: SecurityException) {
            Log.e(TAG, "Security exception getting last known location: ${e.message}")
        }
        
        return bestLocation
    }

    private fun hasLocationPermission(): Boolean {
        val fineLocation = ActivityCompat.checkSelfPermission(
            reactContext,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        
        val coarseLocation = ActivityCompat.checkSelfPermission(
            reactContext,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        
        return fineLocation || coarseLocation
    }

    private fun isLocationEnabled(): Boolean {
        return locationManager?.isProviderEnabled(LocationManager.GPS_PROVIDER) == true ||
               locationManager?.isProviderEnabled(LocationManager.NETWORK_PROVIDER) == true
    }

    @ReactMethod
    fun stopLocationService() {
        Log.d(TAG, "Stopping location service")
        cleanup()
    }

    @ReactMethod
    fun isLocationPermissionGranted(promise: Promise) {
        promise.resolve(hasLocationPermission())
    }

    @ReactMethod
    fun isLocationServicesEnabled(promise: Promise) {
        locationManager = reactContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        promise.resolve(isLocationEnabled())
    }

    override fun onLocationChanged(location: Location) {
        // This method is not used in our implementation
    }

    override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {}
    override fun onProviderEnabled(provider: String) {}
    override fun onProviderDisabled(provider: String) {}

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        cleanup()
    }
}

File: ios/BandMate/GetCurrentLocation.swift
import Foundation
import CoreLocation
import React

@objc(GetCurrentLocation)
class GetCurrentLocation: NSObject, RCTBridgeModule, CLLocationManagerDelegate {
    
    static func moduleName() -> String! {
        return "GetCurrentLocation"
    }
    
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc var bridge: RCTBridge!
    
    private var locationManager: CLLocationManager?
    private var isLocationRequestInProgress = false
    private var currentPromise: RCTPromiseResolveBlock?
    private var currentReject: RCTPromiseRejectBlock?
    private var timeoutTimer: Timer?
    
    private let timeoutDuration: TimeInterval = 15.0 // 15 seconds
    private let lastKnownLocationMaxAge: TimeInterval = 5 * 60 // 5 minutes
    
    override init() {
        super.init()
        setupLocationManager()
    }
    
    private func setupLocationManager() {
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.desiredAccuracy = kCLLocationAccuracyBest
        locationManager?.distanceFilter = 0 // Get immediate updates
    }
    
    @objc
    func getCurrentLocation(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        // Prevent multiple simultaneous requests
        if isLocationRequestInProgress {
            reject("LOCATION_REQUEST_IN_PROGRESS", "A location request is already in progress", nil)
            return
        }
        
        // Check location authorization
        let authorizationStatus = locationManager?.authorizationStatus ?? .notDetermined
        
        switch authorizationStatus {
        case .notDetermined:
            // Request permission first
            currentPromise = resolve
            currentReject = reject
            locationManager?.requestWhenInUseAuthorization()
            return
            
        case .denied, .restricted:
            reject("PERMISSION_DENIED", "Location permission not granted", nil)
            return
            
        case .authorizedWhenInUse, .authorizedAlways:
            break
            
        @unknown default:
            reject("PERMISSION_DENIED", "Unknown location permission status", nil)
            return
        }
        
        // Check if location services are enabled
        if !CLLocationManager.locationServicesEnabled() {
            reject("LOCATION_DISABLED", "Location services are disabled", nil)
            return
        }
        
        currentPromise = resolve
        currentReject = reject
        startLocationRequest()
    }
    
    private func startLocationRequest() {
        isLocationRequestInProgress = true
        
        // Set timeout
        timeoutTimer = Timer.scheduledTimer(withTimeInterval: timeoutDuration, repeats: false) { [weak self] _ in
            self?.handleTimeout()
        }
        
        // First try to get last known location if it's recent enough
        if let lastLocation = locationManager?.location {
            let locationAge = Date().timeIntervalSince(lastLocation.timestamp)
            if locationAge <= lastKnownLocationMaxAge {
                print("GetCurrentLocation: Using last known location (age: \(locationAge)s)")
                handleLocationResult(lastLocation)
                return
            }
        }
        
        // Request new location
        locationManager?.requestLocation()
    }
    
    private func handleLocationResult(_ location: CLLocation) {
        guard isLocationRequestInProgress, let resolve = currentPromise else {
            return
        }
        
        print("GetCurrentLocation: Location result - lat: \(location.coordinate.latitude), lon: \(location.coordinate.longitude), accuracy: \(location.horizontalAccuracy)")
        
        let locationData: [String: Any] = [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            "accuracy": location.horizontalAccuracy,
            "altitude": location.altitude,
            "speed": location.speed >= 0 ? location.speed : 0,
            "heading": location.course >= 0 ? location.course : 0,
            "timestamp": location.timestamp.timeIntervalSince1970 * 1000, // Convert to milliseconds
            "provider": "ios"
        ]
        
        resolve(locationData)
        
        // Emit event for any listeners (optional - for future use)
        // Note: Event emission is not required for basic location functionality
        
        cleanup()
    }
    
    private func handleTimeout() {
        guard isLocationRequestInProgress, let reject = currentReject else {
            return
        }
        
        print("GetCurrentLocation: Location request timed out")
        
        // Try one more time with last known location if available
        if let lastLocation = locationManager?.location {
            print("GetCurrentLocation: Using stale last known location due to timeout")
            handleLocationResult(lastLocation)
            return
        }
        
        reject("TIMEOUT", "Location request timed out", nil)
        cleanup()
    }
    
    private func cleanup() {
        isLocationRequestInProgress = false
        currentPromise = nil
        currentReject = nil
        
        timeoutTimer?.invalidate()
        timeoutTimer = nil
    }
    
    // MARK: - CLLocationManagerDelegate
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        handleLocationResult(location)
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        guard isLocationRequestInProgress, let reject = currentReject else {
            return
        }
        
        print("GetCurrentLocation: Location error - \(error.localizedDescription)")
        
        let errorCode: String
        let errorMessage: String
        
        if let clError = error as? CLError {
            switch clError.code {
            case .denied:
                errorCode = "PERMISSION_DENIED"
                errorMessage = "Location permission was denied. Please enable location services for this app."
            case .locationUnknown:
                errorCode = "LOCATION_ERROR"
                errorMessage = "Location services are not available. Please enable location services."
            case .network:
                errorCode = "LOCATION_ERROR"
                errorMessage = "Network error occurred while getting location."
            default:
                errorCode = "LOCATION_ERROR"
                errorMessage = "Unable to get your location. Please try again."
            }
        } else {
            errorCode = "LOCATION_ERROR"
            errorMessage = error.localizedDescription
        }
        
        reject(errorCode, errorMessage, error)
        cleanup()
    }
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        // Handle authorization changes
        switch status {
        case .authorizedWhenInUse, .authorizedAlways:
            if isLocationRequestInProgress {
                startLocationRequest()
            }
        case .denied, .restricted:
            if let reject = currentReject {
                reject("PERMISSION_DENIED", "Location permission was denied", nil)
                cleanup()
            }
        case .notDetermined:
            // Still waiting for user decision
            break
        @unknown default:
            if let reject = currentReject {
                reject("PERMISSION_DENIED", "Unknown location permission status", nil)
                cleanup()
            }
        }
    }
    
    // MARK: - Additional Methods
    
    @objc
    func stopLocationService(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        print("GetCurrentLocation: Stopping location service")
        cleanup()
        resolve(nil)
    }
    
    @objc
    func isLocationPermissionGranted(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        let status = locationManager?.authorizationStatus ?? .notDetermined
        let isGranted = status == .authorizedWhenInUse || status == .authorizedAlways
        resolve(isGranted)
    }
    
    @objc
    func isLocationServicesEnabled(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(CLLocationManager.locationServicesEnabled())
    }
}